### PR TITLE
FIX: scipy 1d indexing tripped up numpy?

### PR DESCRIPTION
**Pull request by MridulS**
Our nightlies are broken right now https://github.com/networkx/networkx/actions/workflows/nightly-release-test.yml and my hunch is something happened after https://github.com/scipy/scipy/pull/20120

I am not sure if this is expected behavior @dschult ?

### CODE REVIEW

**Comment by rossbar** [comment]
Just to add... I looked at this a bit this morning but I want able to reproduce locally with the scipy nightly wheels. Note that it's only the macos runs that are failing, though at first glance I wouldn't expect this to be platform dependent!

**Comment by MridulS** [comment]
>  I looked at this a bit this morning but I want able to reproduce locally with the scipy nightly wheels.

I can reproduce the error with the wheel, but didn't go bisect and build scipy to find the exact commit 😅

**Comment by rossbar** [comment]
> I can reproduce the error with the wheel, but didn't go bisect and build scipy to find the exact commit 😅

Are you on macos? I can't reproduce the error with the wheel on linux!:

```bash
$ python --version
3.12.3
$ python -m venv nx-devdeps
$ source nx-devdeps/bin/activate
$ pip install -U --pre --extra-index-url https://pypi.org/simple -i https://pypi.anaconda.org/scientific-python-nightly-wheels/simple numpy scipy matplotlib
$ pip install -r requirements/test.txt
$ pip install -e .
$ pip list
<see details below>
$ pytest networkx/algorithms/centrality/tests/test_current_flow_betweenness_centrality.py
```

<details>
<summary>Result of pip list</summary>
<pre>
Package         Version                   Editable project location
--------------- ------------------------- ----------------------------
contourpy       1.3.0.dev1
coverage        7.5.4
cycler          0.12.1
fonttools       4.53.0
iniconfig       2.0.0
kiwisolver      1.4.5
matplotlib      3.10.0.dev344+g651e9109e9
networkx        3.4rc0.dev0               /home/rossbar/repos/networkx
numpy           2.1.0.dev0
packaging       24.1
pillow          10.4.0
pip             24.0
pluggy          1.5.0
pyparsing       3.1.2
pytest          8.2.2
pytest-cov      5.0.0
python-dateutil 2.9.0.post0
scipy           1.15.0.dev0
six             1.16.0
</pre>
</details>

This completes without any errors or warnings for me. If this failure really is platform dependent, then that suggests a problem upstream in scipy - there shouldn't be behavior differences across platforms! Perhaps the platform-specific wheels are out-of-sync?

Either way, I think we should chase down whether there is really a change upstream before patching NX!

**Comment by Peiffap** [comment]
This is pretty strange... I'm on macOS 12.7.5 for reference, and I get the test failures. Here's what I got when trying the commands you gave.

<details>
<summary> Output </summary>

```sh
networkx on  main [!?] via  v3.12.4
❯ python -m venv nx-devdeps

networkx on  main [!?] via  v3.12.4
❯ source nx-devdeps/bin/activate

networkx on  main [!?] via  v3.12.4 (nx-devdeps)
❯ pip install --upgrade pip
Requirement already satisfied: pip in ./nx-devdeps/lib/python3.12/site-packages (24.0)
Collecting pip
  Using cached pip-24.1.1-py3-none-any.whl.metadata (3.6 kB)
Using cached pip-24.1.1-py3-none-any.whl (1.8 MB)
Installing collected packages: pip
  Attempting uninstall: pip
    Found existing installation: pip 24.0
    Uninstalling pip-24.0:
      Successfully uninstalled pip-24.0
Successfully installed pip-24.1.1

networkx on  main [!?] via  v3.12.4 (nx-devdeps) took 3s
❯ pip install -U --pre --extra-index-url https://pypi.org/simple -i https://pypi.anaconda.org/scientific-python-nightly-wheels/simple numpy scipy matplotlib

Looking in indexes: https://pypi.anaconda.org/scientific-python-nightly-wheels/simple, https://pypi.org/simple
Collecting numpy
  Downloading https://pypi.anaconda.org/scientific-python-nightly-wheels/simple/numpy/2.1.0.dev0/numpy-2.1.0.dev0-cp312-cp312-macosx_10_9_x86_64.whl (20.9 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 20.9/20.9 MB 4.5 MB/s eta 0:00:00
WARNING: Cache entry deserialization failed, entry ignored
Collecting scipy
  Downloading https://pypi.anaconda.org/scientific-python-nightly-wheels/simple/scipy/1.15.0.dev0/scipy-1.15.0.dev0-cp312-cp312-macosx_10_13_x86_64.whl (39.1 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 39.1/39.1 MB 7.2 MB/s eta 0:00:00
WARNING: Cache entry deserialization failed, entry ignored
Collecting matplotlib
  Downloading https://pypi.anaconda.org/scientific-python-nightly-wheels/simple/matplotlib/3.10.0.dev344%2Bg651e9109e9/matplotlib-3.10.0.dev344%2Bg651e9109e9-cp312-cp312-macosx_10_12_x86_64.whl (7.9 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.9/7.9 MB 8.5 MB/s eta 0:00:00
WARNING: Cache entry deserialization failed, entry ignored
Collecting contourpy>=1.0.1 (from matplotlib)
  Downloading https://pypi.anaconda.org/scientific-python-nightly-wheels/simple/contourpy/1.3.0.dev1/contourpy-1.3.0.dev1-cp312-cp312-macosx_10_9_x86_64.whl (267 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 267.8/267.8 kB 7.3 MB/s eta 0:00:00
Collecting cycler>=0.10 (from matplotlib)
  Using cached cycler-0.12.1-py3-none-any.whl.metadata (3.8 kB)
Collecting fonttools>=4.22.0 (from matplotlib)
  Using cached fonttools-4.53.0-cp312-cp312-macosx_10_9_universal2.whl.metadata (162 kB)
Collecting kiwisolver>=1.3.1 (from matplotlib)
  Using cached kiwisolver-1.4.5-cp312-cp312-macosx_10_9_x86_64.whl.metadata (6.4 kB)
Collecting packaging>=20.0 (from matplotlib)
  Using cached packaging-24.1-py3-none-any.whl.metadata (3.2 kB)
Collecting pillow>=8 (from matplotlib)
  Using cached pillow-10.4.0-cp312-cp312-macosx_10_10_x86_64.whl.metadata (9.2 kB)
Collecting pyparsing>=2.3.1 (from matplotlib)
  Using cached pyparsing-3.1.2-py3-none-any.whl.metadata (5.1 kB)
Collecting python-dateutil>=2.7 (from matplotlib)
  Using cached python_dateutil-2.9.0.post0-py2.py3-none-any.whl.metadata (8.4 kB)
Collecting six>=1.5 (from python-dateutil>=2.7->matplotlib)
  Using cached six-1.16.0-py2.py3-none-any.whl.metadata (1.8 kB)
Using cached cycler-0.12.1-py3-none-any.whl (8.3 kB)
Using cached fonttools-4.53.0-cp312-cp312-macosx_10_9_universal2.whl (2.8 MB)
Using cached kiwisolver-1.4.5-cp312-cp312-macosx_10_9_x86_64.whl (67 kB)
Using cached packaging-24.1-py3-none-any.whl (53 kB)
Using cached pillow-10.4.0-cp312-cp312-macosx_10_10_x86_64.whl (3.5 MB)
Using cached pyparsing-3.1.2-py3-none-any.whl (103 kB)
Using cached python_dateutil-2.9.0.post0-py2.py3-none-any.whl (229 kB)
Using cached six-1.16.0-py2.py3-none-any.whl (11 kB)
Installing collected packages: six, pyparsing, pillow, packaging, numpy, kiwisolver, fonttools, cycler, scipy, python-dateutil, contourpy, matplotlib
Successfully installed contourpy-1.3.0.dev1 cycler-0.12.1 fonttools-4.53.0 kiwisolver-1.4.5 matplotlib-3.10.0.dev344+g651e9109e9 numpy-2.1.0.dev0 packaging-24.1 pillow-10.4.0 pyparsing-3.1.2 python-dateutil-2.9.0.post0 scipy-1.15.0.dev0 six-1.16.0

networkx on  main [!?] via  v3.12.4 (nx-devdeps) took 45s
❯ pip install -r requirements/test.txt
Collecting pytest>=7.2 (from -r requirements/test.txt (line 3))
  Using cached pytest-8.2.2-py3-none-any.whl.metadata (7.6 kB)
Collecting pytest-cov>=4.0 (from -r requirements/test.txt (line 4))
  Using cached pytest_cov-5.0.0-py3-none-any.whl.metadata (27 kB)
Collecting iniconfig (from pytest>=7.2->-r requirements/test.txt (line 3))
  Using cached iniconfig-2.0.0-py3-none-any.whl.metadata (2.6 kB)
Requirement already satisfied: packaging in ./nx-devdeps/lib/python3.12/site-packages (from pytest>=7.2->-r requirements/test.txt (line 3)) (24.1)
Collecting pluggy<2.0,>=1.5 (from pytest>=7.2->-r requirements/test.txt (line 3))
  Using cached pluggy-1.5.0-py3-none-any.whl.metadata (4.8 kB)
Collecting coverage>=5.2.1 (from coverage[toml]>=5.2.1->pytest-cov>=4.0->-r requirements/test.txt (line 4))
  Using cached coverage-7.5.4-cp312-cp312-macosx_10_9_x86_64.whl.metadata (8.2 kB)
Using cached pytest-8.2.2-py3-none-any.whl (339 kB)
Using cached pytest_cov-5.0.0-py3-none-any.whl (21 kB)
Using cached coverage-7.5.4-cp312-cp312-macosx_10_9_x86_64.whl (205 kB)
Using cached pluggy-1.5.0-py3-none-any.whl (20 kB)
Using cached iniconfig-2.0.0-py3-none-any.whl (5.9 kB)
Installing collected packages: pluggy, iniconfig, coverage, pytest, pytest-cov
Successfully installed coverage-7.5.4 iniconfig-2.0.0 pluggy-1.5.0 pytest-8.2.2 pytest-cov-5.0.0

networkx on  main [!?] via  v3.12.4 (nx-devdeps) took 3s
❯ pip install -e .
Obtaining file:///Users/admin/Documents/foss/networkx
  Installing build dependencies ... done
  Checking if build backend supports build_editable ... done
  Getting requirements to build editable ... done
  Preparing editable metadata (pyproject.toml) ... done
Building wheels for collected packages: networkx
  Building editable for networkx (pyproject.toml) ... done
  Created wheel for networkx: filename=networkx-3.4rc0.dev0-0.editable-py3-none-any.whl size=5995 sha256=0af4a87bfb331a19366080e761136fe6dedf4a892e8961783371d031749bd274
  Stored in directory: /private/var/folders/yr/38162y1s2qx3lvcqgh1k9_p00000gn/T/pip-ephem-wheel-cache-0qrngqpk/wheels/9b/d5/ab/ee13bb19f70a858ef59738b9a81521cc40e52d8757750c3f6a
Successfully built networkx
Installing collected packages: networkx
Successfully installed networkx-3.4rc0.dev0

networkx on  main [!?] via  v3.12.4 (nx-devdeps) took 9s
❯ pip list
Package         Version                   Editable project location
--------------- ------------------------- ------------------------------------
contourpy       1.3.0.dev1
coverage        7.5.4
cycler          0.12.1
fonttools       4.53.0
iniconfig       2.0.0
kiwisolver      1.4.5
matplotlib      3.10.0.dev344+g651e9109e9
networkx        3.4rc0.dev0               /Users/admin/Documents/foss/networkx
numpy           2.1.0.dev0
packaging       24.1
pillow          10.4.0
pip             24.1.1
pluggy          1.5.0
pyparsing       3.1.2
pytest          8.2.2
pytest-cov      5.0.0
python-dateutil 2.9.0.post0
scipy           1.15.0.dev0
six             1.16.0

networkx on  main [!?] via  v3.12.4 (nx-devdeps)
❯ pytest networkx/algorithms/centrality/tests/test_current_flow_betweenness_centrality.py
=================================================== test session starts ====================================================
platform darwin -- Python 3.12.4, pytest-8.2.2, pluggy-1.5.0
rootdir: /Users/admin/Documents/foss/networkx
configfile: pyproject.toml
plugins: cov-5.0.0
collected 20 items

networkx/algorithms/centrality/tests/test_current_flow_betweenness_centrality.py FFFFFFFFFFFFFFFFF...
```

</details>

It'll be hard to see without terminal highlighting, but the nightly installs all printed `WARNING: Cache entry deserialization failed, entry ignored`. I don't think it's related to our problem here, but I thought I'd point it out. @rossbar perhaps it's a Python version thing? Do you happen to have a way of trying on 3.12.4?

**Comment by dschult** [comment]
While I did not expect this result to be backward incompatible, and I haven't dived into what format of sparse array we are using, it does make sense that the new iterating functionality in sparse arrays is backward incompatible.  It used to yield 2D row arrays and now yields 1D arrays. 

I think this means that iterating over sparse arrays has been incorrectly returning 2D arrays for quite a few versions. But it should return 1D objects to match array indexing. This will need to be fixed or reconciled or deprecated or something in scipy. 

Is there something we can do here to avoid the issue? Yes:
Changing `x, y = np.nonzero(row)` to `y = np.nonzero(row)[-1]` should do it since we don't use `x` anywhere here. But there may be other similar code that needs correcting in other ways. 

Thanks for flagging this @MridulS !

**Comment by dschult** [comment]
And I should add that I don't understand why it seems to be failing on some platforms and not others. The change in iteration results of scipy sparse `csr_array` from 2D to 1D should be the same across platforms.

File: networkx/algorithms/centrality/flow_matrix.py [hunk 1]
```diff
@@ -71,7 +71,11 @@ def width(self, L):
         m = 0
         for i, row in enumerate(L):
             w = 0
-            x, y = np.nonzero(row)
+            # scipy.sparse indexing for 1D has changed https://github.com/scipy/scipy/pull/20120
+            try:
+                x, y = np.nonzero(row)
+            except ValueError:
+                y = np.nonzero(row)[0]
```
Comment (dschult) [inline networkx/algorithms/centrality/flow_matrix.py:78]: The subscript `[-1]` will work for old and new versions of scipy
```suggestion
                y = np.nonzero(row)[-1]
```

Comment (dschult) [inline networkx/algorithms/centrality/flow_matrix.py:78]: The index `[-1]` will work for old and new versions of scipy
```suggestion
            y = np.nonzero(row)[-1]
```

**Review by dschult — APPROVED** [review]
Thanks!!

**Comment by rossbar** [comment]
> And I should add that I don't understand why this code seems to be failing on some platforms and not others. The change in iteration results of scipy sparse csr_array from 2D to 1D should be the same across platforms.

My hypothesis is that the nightly wheels for the different platforms are not in sync. For example, maybe the nightly wheels for macos are several commits ahead (or behind) the linux wheels. I haven't actually checked/tested, just a hunch :)

It's probably worth bisecting to see exactly which commit in scipy introduces the behavior to help answer the "how should this be handled in scipy" question.

**Pull request merged**